File: algorithms/526. Beautiful Arrangement/solution.py
class Solution(object):
    def countArrangement(self, N):
        """
        :type N: int
        :rtype: int
        """
        def backtracking(i, remains):
            if not remains:
                return 1
            if remains in cache:
                return cache[remains]
            cnt = 0
            for idx, num in enumerate(remains):
                if num % i == 0 or i % num == 0:
                    cnt += backtracking(i - 1, remains[:idx] + remains[idx + 1:])
            return cache.setdefault(remains, cnt)

        cache = {}
        return backtracking(N, tuple(xrange(1, N+1)))

# 15 / 15 test cases passed.
# Status: Accepted
# Runtime: 116 ms
# beats 82.28%

File: algorithms/526. Beautiful Arrangement/solution.cc
class Solution {
public:
    int countArrangement(int N) {
        vector<int> arr(N);
        for ( int i = 0; i < N; ++i ) {
            arr[i] = i + 1;
        }
        return countArrangement(N, arr);
    }

    int countArrangement(int n, vector<int> & arr) {
        if ( n <= 0 ) {
            return 1;
        }
        int count = 0;
        for ( int i = 0; i < n; ++i ) {
            if ( arr[i] % n == 0 || n % arr[i] == 0 ) {
                swap(arr[i], arr[n-1]);
                count += countArrangement(n - 1, arr);
                swap(arr[n-1], arr[i]);
            }
        }
        return count;
    }
};

// 15 / 15 test cases passed.
// Status: Accepted
// Runtime: 9 ms
// beats 77.61%
